src/attribute.rs

```diff
@@ -1,148 +1,208 @@
+    pub const fn new(string: &'a str) -> Self { Self { string, index: 0 } }
 }
 
-impl FromStr for AttributePairs {
-    type Err = Error;
-
-    fn from_str(input: &str) -> Result<Self, Self::Err> {
-        let mut result = Self::new();
-
-        for line in split(input, ',') {
-            let pair = split(line.trim(), '=');
+impl<'a> Iterator for AttributePairs<'a> {
+    type Item = (&'a str, &'a str);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // return `None`, if there are no more bytes
+        self.string.as_bytes().get(self.index + 1)?;
+
+        let key = {
+            // the position in the string:
+            let start = self.index;
+            // the key ends at an `=`:
+            let end = self
+                .string
+                .char_indices()
+                .skip_while(|(i, _)| *i < self.index)
+                .find_map(|(i, c)| if c == '=' { Some(i) } else { None })?;
+
+            // advance the index to the char after the end of the key (to skip the `=`)
+            // NOTE: it is okay to add 1 to the index, because an `=` is exactly 1 byte.
+            self.index = end + 1;
+
+            ::core::str::from_utf8(&self.string.as_bytes()[start..end]).unwrap()
+        };
+
+        let value = {
+            let start = self.index;
+
+            // find the end of the value by searching for `,`.
+            // it should ignore `,` that are inside double quotes.
+            let mut inside_quotes = false;
+
+            let end = {
+                let mut result = self.string.len();
+
+                for (i, c) in self
+                    .string
+                    .char_indices()
+                    .skip_while(|(i, _)| *i < self.index)
+                {
+                    // if a quote is encountered
+                    if c == '"' {
+                        // update variable
+                        inside_quotes = !inside_quotes;
+                    // terminate if a comma is encountered, which is not in a
+                    // quote
+                    } else if c == ',' && !inside_quotes {
+                        // move the index past the comma
+                        self.index += 1;
+                        // the result is the index of the comma (comma is not included in the
+                        // resulting string)
+                        result = i;
+                        break;
+                    }
+                }
 
-            if pair.len() < 2 {
-                continue;
-            }
+                result
+            };
 
-            let key = pair[0].trim().to_uppercase();
-            let value = pair[1].trim().to_string();
-            if value.is_empty() {
-                continue;
-            }
+            self.index += end;
+            self.index -= start;
 
-            result.insert(key.trim().to_string(), value.trim().to_string());
-        }
+            ::core::str::from_utf8(&self.string.as_bytes()[start..end]).unwrap()
```

Comment:
It should be possible to check if the indecies are at a boundery to make sure that this will not panic. for example with: https://doc.rust-lang.org/std/primitive.str.html#method.is_char_boundary

Reply:
This should never panic, therefore I made it an `unwrap` instead of returning an error or using the unsafe equivalent `from_utf8_unchecked`.

This is ensured, because only indexes returned by [`str::char_indices`](https://doc.rust-lang.org/std/primitive.str.html#method.char_indices) are assigned to `self.index`.

The full function:

```rust
fn next(&mut self) -> Option<Self::Item> {
    // return `None`, if there are no more bytes
    self.string.as_bytes().get(self.index + 1)?;

    let key = {
        // the position in the string:
        let start = self.index;
        // the key ends at an `=`:
        let end = self
            .string
            .char_indices()
            .skip_while(|(i, _)| *i < self.index)
            .find_map(|(i, c)| if c == '=' { Some(i) } else { None })?;

        // advance the index to the char after the end of the key (to skip the `=`)
        // NOTE: it is okay to add 1 to the index, because an `=` is exactly 1 byte.
        self.index = end + 1;

        ::core::str::from_utf8(&self.string.as_bytes()[start..end]).unwrap()
    };

    let value = {
        let start = self.index;

        // find the end of the value by searching for `,`.
        // it should ignore `,` that are inside double quotes.
        let mut inside_quotes = false;

        let end = {
            let mut result = self.string.len();

            for (i, c) in self
                .string
                .char_indices()
                .skip_while(|(i, _)| *i < self.index)
            {
                // if a quote is encountered
                if c == '"' {
                    // update variable
                    inside_quotes = !inside_quotes;
                // terminate if a comma is encountered, which is not in a
                // quote
                } else if c == ',' && !inside_quotes {
                    // move the index past the comma
                    self.index += 1;
                    // the result is the index of the comma (comma is not included in the
                    // resulting string)
                    result = i;
                    break;
                }
            }

            result
        };

        self.index += end;
        self.index -= start;

        ::core::str::from_utf8(&self.string.as_bytes()[start..end]).unwrap()
    };

    Some((key.trim(), value.trim()))
}
```

Reply:
Yeah that makes sense, I will have another look through it when I have some time next week if it is not merged before then.